### PR TITLE
Draft: Diffusers for Mac

### PR DESCRIPTION
**Pull request by pcuenca**

### CODE REVIEW

File: fast-mac-diffusion.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+---
+title: Fast Stable Diffusion on your Mac using Swift Diffusers
+thumbnail: /blog/assets/fast-mac-diffusion/thumbnail.png
+authors:
+- user: pcuenq
+- user: reach-vb
+---
+
+# Fast Stable Diffusion with Diffusers for Mac
```
Comment (pcuenca) [inline fast-mac-diffusion.md:9]: Unsure about the title, that's why it still doesn't match in the metadata.

File: fast-mac-diffusion.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+<!-- {blog_metadata} -->
+<!-- {authors} -->
+
+Diffusers for Mac is a native app to generate images from a text description of what you want. It uses state-of-the-art diffusion models, like Stable Diffusion, contributed by the community to the Hugging Face Hub, and converted to Core ML for maximum performance. We have just released version 1.1 of the app in the Mac App Store, with significant performance improvements, a lot of UI tweaks and some bug fixes. We think it's a solid foundation on which to build new features. And it's [fully open source](https://github.com/huggingface/swift-coreml-diffusers) with a [permissive license](https://github.com/huggingface/swift-coreml-diffusers/blob/main/LICENSE), so you can build on it too!
+
+**TODO**: screenshot of the new UI
+
+## What exactly is Diffusers for Mac anyway?
```
Comment (pcuenca) [inline fast-mac-diffusion.md:18]: I wanted to introduce the app because there was no previous communication about it (other than my tweet). I tried to include it after the main topic about performance, but I couldn't get it to flow well.

File: fast-mac-diffusion.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+- user: pcuenq
+- user: reach-vb
+---
+
+# Fast Stable Diffusion with Diffusers for Mac
+
+<!-- {blog_metadata} -->
+<!-- {authors} -->
+
+Diffusers for Mac is a native app to generate images from a text description of what you want. It uses state-of-the-art diffusion models, like Stable Diffusion, contributed by the community to the Hugging Face Hub, and converted to Core ML for maximum performance. We have just released version 1.1 of the app in the Mac App Store, with significant performance improvements, a lot of UI tweaks and some bug fixes. We think it's a solid foundation on which to build new features. And it's [fully open source](https://github.com/huggingface/swift-coreml-diffusers) with a [permissive license](https://github.com/huggingface/swift-coreml-diffusers/blob/main/LICENSE), so you can build on it too!
```
Comment (Vaibhavs10) [inline fast-mac-diffusion.md:14]: Suggestion to make it pack a bit more punch:

Transform your text into stunning images with ease using Diffusers for Mac - the native app powered by state-of-the-art diffusion models. It leverages a bouquet of SoTA Text-to-Image models contributed by the community on Hugging Face Hub, and converted to Core ML for blazingly fast performance.  Our latest version, 1.1, is now available on the Mac App Store with significant performance upgrades and user-friendly interface tweaks. You can trust our Solid foundation for future feature updates. Plus, our app is fully open source with a permissive license, so you can build on it too! Check out our GitHub repository at https://github.com/huggingface/swift-coreml-diffusers for more information.

File: fast-mac-diffusion.md (hunk 2 above)
Comment (Vaibhavs10) [reply fast-mac-diffusion.md:18]: That's okay, I think the flow here can be the following:
1. Intro: Mac app for diffusers
2. A bit of deep-dive on it (this should focus on how the community can leverage it to build there own Mac apps and so on)
3. Performance benchmarks
4. Community call for features and benchmarks

WDYT?

File: fast-mac-diffusion.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+**TODO**: screenshot of the new UI
+
+## What exactly is Diffusers for Mac anyway?
+
+The Diffusers app is the Mac counterpart to our [`diffusers` 🧨 library](https://github.com/huggingface/diffusers). This library is written in Python with PyTorch, and uses a modular design to train and run diffusion models. `diffusers` supports many different models and tasks, is highly configurable and very well optimized. It runs on Mac, too, using PyTorch's [`mps` accelerator](https://huggingface.co/docs/diffusers/optimization/mps), which is an alternative to `cuda` on Apple Silicon.
```
Comment (Vaibhavs10) [inline fast-mac-diffusion.md:20]: I'd add the link to the app (app store) as well as GitHub.

File: fast-mac-diffusion.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+Why would you want to run a native Mac app then? There are many reasons:
+- It uses Core ML models, instead of the original PyTorch ones. This is important because they allow for [additional optimizations](link to apple's post on transformers/diffusers) relevant to the specifics of Apple hardware, and because Core ML models can run on all the compute devices in your system: the CPU, the GPU and the Neural Engine, _at once_ – the system will decide what portions of your model to run on each device to make it as fast as possible. PyTorch's `mps` device cannot use the Neural Engine.
```
Comment (Vaibhavs10) [inline fast-mac-diffusion.md:23]: (nit) the system -> app

I wonder if we should add something along the lines of "without the need to set up the environment, etc" somewhere.

File: fast-mac-diffusion.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+- It's local, and private. You don't need credits for online services and won't experience long queues – just generate all the images you want and use them for fun or work. Privacy is guaranteed: your prompts and images are yours to use, and will never leave your computer (unless you choose to share them).
+- [It's open source](https://github.com/huggingface/swift-coreml-diffusers), and it uses the latest languages and technologies for Mac and iOS development. If you are technically inclined, you can use Xcode to extend our Swift code as you like. We welcome your contributions, too!
+
+## Performance Improvements
+
+**TL;DR:** Generation is up to **twice as fast** on Diffusers 1.1, depending on the computer you use. Benchmark and details follow.
+
+The first version of the app we released a few days ago runs models on a combination of CPU and GPU, and does not use the Neural Engine. This is because it produced very good results on our test systems (i.e., my iMac and my laptop), it's most compatible, and requires less downloads. Use of the Neural Engine, or ANE, requires a different version of the Core ML model.
```
Comment (Vaibhavs10) [inline fast-mac-diffusion.md:32]: I'd remove this. Let's not mention the past version here.
I think we should focus on the performance benchmarks here.

**Comment by Vaibhavs10** [comment]
I think it is in a fantastic shape, just a suggestion on the overall structure:

1. Intro: Mac app for diffusers
2. A bit of deep-dive on it (this should focus on how the community can leverage it to build there own Mac apps and so on)
3. Performance benchmarks
4. Community call for features and benchmarks

We can work on the overall grammar and spell checks towards the end.

File: fast-mac-diffusion.md (hunk 1 above)
Comment (pcuenca) [reply fast-mac-diffusion.md:9]: Thought about it too. Not sure, I suspect non-developers would be interested in it too so it sounds safer to use "Mac" rather than "Swift" (but mention Swift in the text)

File: fast-mac-diffusion.md (hunk 5 above)
Comment (pcuenca) [reply fast-mac-diffusion.md:23]: > (nit) the system -> app

It's not the app, it's the Core ML framework (or "the OS", in general).

File: fast-mac-diffusion.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+Since we released we've done a lot more testing on our personal systems, and on additional Macs we rented from cloud services. These are the results of our benchmark:
+
+**TODO** Table.
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply fast-mac-diffusion.md:36]: Not sure if we should add PyTorch, it may add confusion.

There's also the new setting to use CPU+GPU+ANE, we could include those figures too or just state qualitatively that it doesn't help in the systems we tested (I'd do the latter for simplicity).

File: fast-mac-diffusion.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+The amount of memory does not seem to play a big factor on performance, but the CPU performance (i.e., the number of _performance cores_ in the system) does. My M1 Max laptop is a lot faster using the GPU than it is with the ANE. However, standard M1 processors found in Mac Minis are **twice as fast** using ANE than GPU. Interestingly, using _both_ the GPU and ANE accelerators does not improve performance with respect to the best results obtained with just one of them.
+
+So for version 1.1 we are now automatically selecting the best accelerator based on the computer where the app runs. However, we don't have access to many other Mac models because they are not offered by any cloud services. For example, we haven't been able to test performance on any of the "Pro" Macbook Pro variants. If you'd like to help us gather data to keep improving the app, read on!
+
+## Community Call for Benchmark Data
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply fast-mac-diffusion.md:42]: Sounds great! Do you have enough ideas for the "good first issues"? :)

I forgot to include the models we are most interested in: M1 Pro, M2 Pro, M2 Max. I'll add the list and we'll decide if we want to keep it here or just mention it in GitHub.

File: fast-mac-diffusion.md (hunk 5 above)
Comment (Vaibhavs10) [reply fast-mac-diffusion.md:23]: Then how about we call it the framework? The system sounds a bit too vague.

File: fast-mac-diffusion.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (Vaibhavs10) [reply fast-mac-diffusion.md:36]: I guess my reason for adding PyTorch was to compare `mps`, I was hoping this helps us move over PyTorch hackers on mac over to the framework.

File: fast-mac-diffusion.md (hunk 2 above)
Comment (pcuenca) [reply fast-mac-diffusion.md:18]: 👍

**Comment by julien-c** [comment]
good idea IMO to do a post about it!

File: fast-mac-diffusion.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (pcuenca) [reply fast-mac-diffusion.md:36]: Yes, but it can be tricky to test. `diffusers` has a lot of optional memory optimizations that also help with performance, especially on memory pressured devices. So finding the right combinations will take some time, and some people may still complain that "this one trick" works better for them :) I'll run some tests though.

File: fast-mac-diffusion.md [hunk 9]
```diff
@@ -11,7 +11,7 @@ authors:
...
+Transform your text into stunning images with ease using Diffusers for Mac, a native app powered by state-of-the-art diffusion models. It leverages a bouquet of SoTA Text-to-Image models contributed by the community to the 🤗 Hub, and converted to Core ML for blazingly fast performance. Our latest version, 1.1, is now available on the [Mac App Store](https://apps.apple.com/app/diffusers/id1666309574) with significant performance upgrades and user-friendly interface tweaks. It's a solid foundation for future feature updates. Plus, the app is fully open source with a permissive [license](https://github.com/huggingface/swift-coreml-diffusers/blob/main/LICENSE), so you can build on it too! Check out our GitHub repository at https://github.com/huggingface/swift-coreml-diffusers for more information.
```
Comment (pcuenca) [inline fast-mac-diffusion.md:14]: @Vaibhavs10 I think we don't use the emoji with the Hub, it's usually Hugging Face Hub on first mention and then just "the Hub".

File: fast-mac-diffusion.md [hunk 10]
```diff
@@ -11,18 +11,18 @@ authors:
...
+The 🧨Diffusers app ([App Store](https://apps.apple.com/app/diffusers/id1666309574), [source code](https://github.com/huggingface/swift-coreml-diffusers)) is the Mac counterpart to our [🧨`diffusers` library](https://github.com/huggingface/diffusers). This library is written in Python with PyTorch, and uses a modular design to train and run diffusion models. 🧨`diffusers` supports many different models and tasks, and are highly configurable and well optimized. It runs on Mac, too, using PyTorch's [`mps` accelerator](https://huggingface.co/docs/diffusers/optimization/mps), which is an alternative to `cuda` on Apple Silicon.
```
Comment (pcuenca) [inline fast-mac-diffusion.md:20]: We usually refer to diffusers in singular, as it's a library. So it would be `is highly configurable...` (and matches `supports` earlier in the same line)

File: fast-mac-diffusion.md [hunk 11]
```diff
@@ -27,11 +27,11 @@ Why would you want to run a native Mac app then? There are many reasons:
 
 ## Performance Benchmarks
 
-**TL;DR:** Generation is up to **twice as fast** on Diffusers 1.1, depending on the computer you use. Benchmark and details follow.
+**TL;DR:** Depending on your computer Text-to-Image Generation can be up to **twice as fast** on Diffusers 1.1. ⚡️
 
-We've done a lot of testing on our own systems and on Macs we rented from cloud services to determine the best combinations of compute devices that yield optimum performance. For some computers it's best to use the GPU, while others work better when the Neural Engine, or ANE, is engaged.
+We've done a lot of testing on Macs to determine the best combinations of compute devices that yield optimum performance. For some computers it's best to use the GPU, while others work better when the Neural Engine, or ANE, is engaged.
```
Comment (pcuenca) [inline fast-mac-diffusion.md:32]:
```suggestion
We've done a lot of testing on several Macs to determine the best combinations of compute devices that yield optimum performance. For some computers it's best to use the GPU, while others work better when the Neural Engine, or ANE, is engaged.
Write
```

File: fast-mac-diffusion.md [hunk 12]
```diff
@@ -27,11 +27,11 @@ Why would you want to run a native Mac app then? There are many reasons:
...
 
-These are the results of our benchmark. All combinations use the CPU in addition to either the GPU or the ANE.
+Come checkout out our benchmarks. All the combinations use the CPU in addition to either the GPU or the ANE.
```
Comment (pcuenca) [inline fast-mac-diffusion.md:34]:
```suggestion
Come checkout our benchmarks. All the combinations use the CPU in addition to either the GPU or the ANE.
```
